balance/src/lib.rs

```diff
@@ -0,0 +1,350 @@
+extern crate futures;
+extern crate h2;
+extern crate http;
+extern crate tower_balance;
+extern crate tower_h2;
+
+use futures::{Async, Poll};
+use std::marker::PhantomData;
+use tower_balance::load::Instrument;
+
+/// Instruments HTTP responses to drop handles when their first body message is received.
+#[derive(Clone, Debug)]
+pub struct PendingUntilFirstData;
+
+/// Instruments HTTP responses to drop handles when their streams completes.
+#[derive(Clone, Debug)]
+pub struct PendingUntilEos;
+
+/// An instrumented HTTP body that drops its handle according to the `S`-typed strategy.
+pub struct Body<T, B, S> {
+    handle: Option<T>,
+    body: B,
+    _p: PhantomData<S>,
+}
+
+// Instrumentation helper.
+fn instrument_body<T, B, S>(handle: T, rsp: http::Response<B>) -> http::Response<Body<T, B, S>>
+where
+    B: tower_h2::Body,
+{
+    let (parts, body) = rsp.into_parts();
+    let handle = if body.is_end_stream() {
+        None
+    } else {
+        Some(handle)
+    };
+    let body = Body {
+        handle,
+        body,
+        _p: PhantomData,
+    };
+    http::Response::from_parts(parts, body)
+}
+
+// ==== PendingUntilEos ====
+
+impl<T, B> Instrument<T, http::Response<B>> for PendingUntilFirstData
+where
+    T: Sync + Send + 'static,
+    B: tower_h2::Body + 'static,
+{
+    type Output = http::Response<Body<T, B, Self>>;
```

Comment:
I wonder if there should be separate body types instead of using a type annotation. Separate types would allow each permutation to evolve over time by adding fields if needed. I also don’t see anything that requires the single body struct. 

I wonder what others think?

Reply:
I had implemented that way initially, but the body types were exactly the same so i anticipated a review asking to change that ;) i'm happy either way.